### PR TITLE
MGMT-2782 - IPv6 Phase 3 changes

### PR DESCRIPTION
**Pull request by ori-amizur**
- Add host networks support for update-cluster and get-cluster
- Add safety checks in advanced networking validations
- Add unit and subsystem tests

/cc @ronniel1 
/cc @empovit

### CODE REVIEW

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -1991,14 +1991,16 @@ func calculateHostNetworks(log logrus.FieldLogger, cluster *common.Cluster) []*m
 			continue
 		}
 		for _, intf := range inventory.Interfaces {
-			for _, ipv4Address := range intf.IPV4Addresses {
-				_, ipnet, err := net.ParseCIDR(ipv4Address)
-				if err != nil {
-					log.WithError(err).Warnf("Could not parse CIDR %s", ipv4Address)
-					continue
+			for _, addresses := range [][]string{intf.IPV4Addresses, intf.IPV6Addresses} {
```
Comment (empovit) [inline internal/bminventory/inventory.go:1994]: Won't it be more intuitive to join the two slices and avoid the nested loop? `for _address := range join(intf.IPV4Addresses, intf.IPV6Addresses...) {`

File: internal/installcfg/installcfg_test.go [hunk 2]
```diff
@@ -33,7 +33,7 @@ var _ = Describe("installcfg", func() {
 			BaseDNSDomain:          "redhat.com",
 			APIVip:                 "102.345.34.34",
 			IngressVip:             "376.5.56.6",
-			InstallConfigOverrides: `{"networking":{"networkType": "OVN-Kubernetes"},"fips":true}`,
+			InstallConfigOverrides: `{"networking":{"networkType": "OVNKubernetes"},"fips":true}`,
```
Comment (empovit) [inline internal/installcfg/installcfg_test.go:36]: Looks like a constant can be helpful here.

File: internal/network/cidr_validations.go [hunk 3]
```diff
@@ -59,6 +59,13 @@ func max(x, y int) int {
...
 }
 
+func min(x, y int) int {
+	if x < y {
+		return x
+	}
+	return y
+}
+
 func VerifyClusterCidrSize(hostNetworkPrefix int, clusterNetworkCIDR string, numberOfHosts int) error {
```
Comment (empovit) [inline internal/network/cidr_validations.go:69]: I couldn't find UT for this function. Probably they're worth adding.

File: subsystem/ipv6_test.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+					"1001:db8::10/120",
+				},
+			},
+		},
+		SystemVendor: &models.SystemVendor{Manufacturer: "manu", ProductName: "prod", SerialNumber: "3534"},
+		Timestamp:    1601853088,
+	}
+)
+
+var _ = Describe("IPv6 installation", func() {
```
Comment (empovit) [inline subsystem/ipv6_test.go:38]: Did you consider parameterizing the IPv4 installation tests so that the same scenarios can run twice - with IPv4 and IPv6?

File: subsystem/ipv6_test.go [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+		},
+		ClusterID: clusterID,
+	})
+	Expect(err).NotTo(HaveOccurred())
+	waitForClusterState(ctx, clusterID, models.ClusterStatusReady, 60*time.Second, clusterReadyStateInfo)
+
+	return hosts
+}
+
+func generateFullMeshConnectivityV6(ctx context.Context, startIPAddress string, hosts ...*models.Host) {
```
Comment (empovit) [inline subsystem/ipv6_test.go:265]: I think we already have a similar function. Couldn't it be reused?

File: subsystem/ipv6_test.go (hunk 4 above)
Comment (ori-amizur) [reply subsystem/ipv6_test.go:38]: Yes, then we will double the duration of the tests

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (ori-amizur) [reply internal/bminventory/inventory.go:1994]: join or append?

File: internal/installcfg/installcfg_test.go (hunk 2 above)
Comment (ori-amizur) [reply internal/installcfg/installcfg_test.go:36]: OK

File: internal/network/cidr_validations.go (hunk 3 above)
Comment (ori-amizur) [reply internal/network/cidr_validations.go:69]: They are part of validations (transitions_test).  I will add some tests

File: subsystem/ipv6_test.go (hunk 5 above)
Comment (ori-amizur) [reply subsystem/ipv6_test.go:265]: Will try

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (empovit) [reply internal/bminventory/inventory.go:1994]: `append` of course, sorry. Updated the comment. I tested it and it works.

File: subsystem/ipv6_test.go (hunk 4 above)
Comment (empovit) [reply subsystem/ipv6_test.go:38]: I didn't mean all of them, just a subset that is likely to depend on IP addresses.

Comment (ori-amizur) [reply subsystem/ipv6_test.go:38]: Almost all

File: subsystem/cluster_test.go [hunk 6]
```diff
@@ -634,8 +634,9 @@ var _ = Describe("cluster update - BaseDNS", func() {
 
 var _ = Describe("cluster install", func() {
 	var (
-		ctx           = context.Background()
-		cluster       *models.Cluster
+		ctx     = context.Background()
+		cluster *models.Cluster
+		//nolint
```
Comment (YuviGold) [inline subsystem/cluster_test.go:639]: Why is it here for?

File: internal/network/cidr_validations.go [hunk 7]
```diff
@@ -70,10 +77,11 @@ func VerifyClusterCidrSize(hostNetworkPrefix int, clusterNetworkCIDR string, num
 		return errors.Errorf("Host prefix, now %d, must be less than or equal to %d to allow at least 128 addresses", hostNetworkPrefix, bits-MinMaskDelta)
 	}
 	requestedNumHosts := max(4, numberOfHosts)
-	possibleNumHosts := uint64(1) << max(hostNetworkPrefix-clusterNetworkPrefix, 0)
+	// 63 to avoid overflow
```
Comment (YuviGold) [inline internal/network/cidr_validations.go:80]: Should have a test

File: subsystem/ipv6_test.go [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+				ServiceNetworkCidr:       &serviceCIDR,
+				SSHPublicKey:             "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQC50TuHS7aYci+U+5PLe/aW/I6maBi9PBDucLje6C6gtArfjy7udWA1DCSIQd+DkHhi57/s+PmvEjzfAfzqo+L+/8/O2l2seR1pPhHDxMR/rSyo/6rZP6KIL8HwFqXHHpDUM4tLXdgwKAe1LxBevLt/yNl8kOiHJESUSl+2QSf8z4SIbo/frDD8OwOvtfKBEG4WCb8zEsEuIPNF/Vo/UxPtS9pPTecEsWKDHR67yFjjamoyLvAzMAJotYgyMoxm8PTyCgEzHk3s3S4iO956d6KVOEJVXnTVhAxrtLuubjskd7N4hVN7h2s4Z584wYLKYhrIBL0EViihOMzY4mH3YE4KZusfIx6oMcggKX9b3NHm0la7cj2zg0r6zjUn6ZCP4gXM99e5q4auc0OEfoSfQwofGi3WmxkG3tEozCB8Zz0wGbi2CzR8zlcF+BNV5I2LESlLzjPY5B4dvv5zjxsYoz94p3rUhKnnPM2zTx1kkilDK5C5fC1k9l/I/r5Qk4ebLQU= oscohen@localhost.localdomain",
+			},
+		})
+		Expect(err).NotTo(HaveOccurred())
+		cluster = registerClusterReply.GetPayload()
+		clusterID = *cluster.ID
+		log.Infof("Register cluster %s", cluster.ID.String())
+	})
+	It("[only_k8s]report_host_progress", func() {
```
Comment (YuviGold) [inline subsystem/ipv6_test.go:67]: This test logic doesn't need to be verified twice (it's been copied from `cluster_test`). You should minimize it to only IPv6 related stuff.

File: internal/host/host_test.go [hunk 9]
```diff
@@ -901,6 +901,10 @@ func masterInventory() string {
 	return masterInventoryWithHostname("master-hostname")
 }
 
+func masterInventoryV6() string {
```
Comment (YuviGold) [inline internal/host/host_test.go:904]: Should we verify also `workerInventoryV6`?

File: internal/network/cidr_validations.go (hunk 7 above)
Comment (ori-amizur) [reply internal/network/cidr_validations.go:80]: OK

File: subsystem/cluster_test.go (hunk 6 above)
Comment (ori-amizur) [reply subsystem/cluster_test.go:639]: Lint started complaining about overriding this value.  I did not touch this part of code at all.  So this tells lint not to complain.

File: subsystem/ipv6_test.go (hunk 8 above)
Comment (ori-amizur) [reply subsystem/ipv6_test.go:67]: What do you mean by "only IPv6 related stuff"?

File: subsystem/cluster_test.go (hunk 6 above)
Comment (YuviGold) [reply subsystem/cluster_test.go:639]: I think it should be resolved instead of disabling the linter.

File: subsystem/ipv6_test.go (hunk 8 above)
Comment (YuviGold) [reply subsystem/ipv6_test.go:67]: This test for example checks various "report" scenarios. It isn't related to IPv4 nor IPv6.
You should just verify an installation or specific edge-cases where you think it might be relevant.

File: subsystem/cluster_test.go (hunk 6 above)
Comment (ori-amizur) [reply subsystem/cluster_test.go:639]: OK

File: subsystem/ipv6_test.go (hunk 8 above)
Comment (ori-amizur) [reply subsystem/ipv6_test.go:67]: Actually,  there are no known edge cases specific to IPv6 that should be tested here.  Just happy flow.

File: internal/host/host_test.go (hunk 9 above)
Comment (ori-amizur) [reply internal/host/host_test.go:904]: If we need.  Right now we don't

File: subsystem/ipv6_test.go (hunk 8 above)
Comment (ori-amizur) [reply subsystem/ipv6_test.go:67]: Left just happy flow

**Review by YuviGold — APPROVED** [review]
LGTM
